ext/strings_test.go

```diff
@@ -329,3 +335,673 @@ func TestStringsWithExtension(t *testing.T) {
+				"dynInf": math.Inf(1),
+			},
+			expectedOutput: "NaN: NaN, infinity: ∞",
+		},
+		{
+			name:       "dyntype support for timestamp",
+			format:     "dyntype timestamp: %s",
+			formatArgs: `dynTime`,
+			dynArgs: map[string]any{
+				"dynTime": time.Date(2009, time.November, 10, 23, 0, 0, 0, time.UTC),
+			},
+			expectedOutput: "dyntype timestamp: 2009-11-10T23:00:00Z",
+		},
+		{
+			name:       "dyntype support for duration",
+			format:     "dyntype duration: %s",
+			formatArgs: `dynDuration`,
+			dynArgs: map[string]any{
+				"dynDuration": mustParseDuration("2h25m47s"),
+			},
+			expectedOutput: "dyntype duration: 8747s",
+		},
+		{
+			name:       "dyntype support for lists",
+			format:     "dyntype list: %s",
+			formatArgs: `dynList`,
+			dynArgs: map[string]any{
+				"dynList": []any{6, 4.2, "a string"},
+			},
+			expectedOutput: `dyntype list: [6, 4.200000, "a string"]`,
+		},
+		{
+			name:       "dyntype support for maps",
+			format:     "dyntype map: %s",
+			formatArgs: `dynMap`,
+			dynArgs: map[string]any{
+				"dynMap": map[any]any{
+					"strKey": "x",
+					true:     42,
+					int64(6): mustParseDuration("7m2s"),
+				},
+			},
+			expectedOutput: `dyntype map: {"strKey":"x", 6:duration("422s"), true:42}`,
+		},
+		{
+			name:       "unrecognized formatting clause",
+			format:     "%a",
+			formatArgs: "1",
+			err:        "could not parse formatting clause: unrecognized formatting clause \"a\"",
+		},
+		{
+			name:       "out of bounds arg index",
+			format:     "%d %d %d",
+			formatArgs: "0, 1",
+			err:        "index 2 out of range",
+		},
+		{
+			name:       "string substitution is not allowed with binary clause",
+			format:     "string is %b",
+			formatArgs: `"abc"`,
+			err:        "error during formatting: only integers and bools can be formatted as binary, was given string",
+		},
+		{
+			name:       "duration substitution not allowed with decimal clause",
+			format:     "%d",
+			formatArgs: `duration("30m2s")`,
+			err:        "error during formatting: decimal clause can only be used on integers, was given google.protobuf.Duration",
+		},
+		{
+			name:       "string substitution not allowed with octal clause",
+			format:     "octal: %o",
+			formatArgs: `"a string"`,
+			err:        "error during formatting: octal clause can only be used on integers, was given string",
+		},
+		{
+			name:       "double substitution not allowed with hex clause",
+			format:     "double is %x",
+			formatArgs: "0.5",
+			err:        "error during formatting: only integers, byte buffers, and strings can be formatted as hex, was given double",
+		},
+		{
+			name:       "uppercase not allowed for scientific clause",
+			format:     "double is %E",
+			formatArgs: "0.5",
+			err:        `could not parse formatting clause: unrecognized formatting clause "E"`,
+		},
+		{
+			name:       "object not allowed",
+			format:     "object is %s",
+			formatArgs: `ext.TestAllTypes{PbVal: test.TestAllTypes{}}`,
```

Comment:
What happens if you have a list of objects or a map containing objects?

Reply:
It fails with an error, which I feel is the correct behavior - I've added a couple of tests to make sure this is made clear though.